### ISSUE TITLE
Support for Java 13

### ISSUE THREAD

**Issue by xtbe** [eclipse-ee4j/jersey#4402]
Hi,

Is it possible to run Jersey on Java 13 (v2.30.1)

Asking because I get this stacktrace when running it on that version. Thanks for the feedback.

HTTP Status 500 – Internal Server Error
Type Exception Report

Message Servlet.init() for servlet [Jersey REST Service] threw exception

Description The server encountered an unexpected condition that prevented it from fulfilling the request.

Exception

javax.servlet.ServletException: Servlet.init() for servlet [Jersey REST Service] threw exception
	org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526)
	org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:861)
	org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1579)
	org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	java.base/java.lang.Thread.run(Unknown Source)
Root Cause

java.lang.IllegalArgumentException
	jersey.repackaged.org.objectweb.asm.ClassReader.<init>(ClassReader.java:171)
	jersey.repackaged.org.objectweb.asm.ClassReader.<init>(ClassReader.java:153)
	jersey.repackaged.org.objectweb.asm.ClassReader.<init>(ClassReader.java:425)
	org.glassfish.jersey.server.internal.scanning.AnnotationAcceptingListener.process(AnnotationAcceptingListener.java:146)
	org.glassfish.jersey.server.ResourceConfig.scanClasses(ResourceConfig.java:885)
	org.glassfish.jersey.server.ResourceConfig._getClasses(ResourceConfig.java:839)
	org.glassfish.jersey.server.ResourceConfig.getClasses(ResourceConfig.java:748)
	org.glassfish.jersey.server.ResourceConfig$RuntimeConfig.<init>(ResourceConfig.java:1175)
	org.glassfish.jersey.server.ResourceConfig$RuntimeConfig.<init>(ResourceConfig.java:1148)
	org.glassfish.jersey.server.ResourceConfig.createRuntimeConfig(ResourceConfig.java:1144)
	org.glassfish.jersey.server.ApplicationHandler$RuntimeConfigConfigurator.init(ApplicationHandler.java:181)
	org.glassfish.jersey.server.ApplicationHandler.lambda$initialize$0(ApplicationHandler.java:289)
	org.glassfish.jersey.server.ApplicationHandler$$Lambda$170.0000000000000000.accept(Unknown Source)
	java.base/java.util.Arrays$ArrayList.forEach(Unknown Source)
	org.glassfish.jersey.server.ApplicationHandler.initialize(ApplicationHandler.java:289)
	org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:258)
	org.glassfish.jersey.servlet.WebComponent.<init>(WebComponent.java:311)
	org.glassfish.jersey.servlet.ServletContainer.init(ServletContainer.java:154)
	org.glassfish.jersey.servlet.ServletContainer.init(ServletContainer.java:347)
	javax.servlet.GenericServlet.init(GenericServlet.java:158)
	org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526)
	org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:861)
	org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1579)
	org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	java.base/java.lang.Thread.run(Unknown Source)
Note The full stack trace of the root cause is available in the server logs.

**Comment by garysheppardjr** [comment]
It looks like it depends on what version of Jersey you're using. It seems that something in the last several months should work with Java 13-targeted bytecode. If your Jersey is older, try targeting Java 11 when you compile your code.

I say this based on my experience with Java 8, Java 11, Jersey 1.x, and Jersey 2.x.

- Targeting Java 8, running Java 8, using Jersey 1.x: OK ✔
- Targeting Java 8, running Java 11, using Jersey 1.x: OK ✔
- Targeting Java 11, running Java 11, using Jersey 1.x: fail ❌
- Targeting Java 11, running Java 11, using Jersey 2.x: OK ✔

Why doesn't Java 11-targeted bytecode work with Jersey 1.x? Because of this:

https://github.com/jersey/jersey-1.x/blob/864a01d7be490ab93d2424da3e446ad8eb84b1e8/jersey-server/src/main/java/jersey/repackaged/org/objectweb/asm/ClassReader.java#L168-L171

```
// checks the class version
if (readShort(off + 6) > Opcodes.V1_8) {
    throw new IllegalArgumentException();
}
```

We have since migrated from Jersey 1.x to Jersey 2.x, and now we can target Java 11 when we compile our code.

Oddly enough, the current Jersey source says that everything up to Java 15 should work:

https://github.com/eclipse-ee4j/jersey/blob/584664c81b6983c2118f67b7f2d8db237cfee323/core-server/src/main/java/jersey/repackaged/org/objectweb/asm/ClassReader.java#L199-L202

It looks like Java 13 was enabled in #4095. I would think that any Jersey release that includes that commit should work with Java 13-targeted bytecode.

**Comment by jansupol** [comment]
Jersey 2.30.1 works fine on JDK 13, I suppose, at least we do not have any issue on our test environment and Travis.

My suspicion is that you have an old version of ClassReader, as the stacktrace
> jersey.repackaged.org.objectweb.asm.ClassReader.(ClassReader.java:171)
jersey.repackaged.org.objectweb.asm.ClassReader.(ClassReader.java:153)
jersey.repackaged.org.objectweb.asm.ClassReader.(ClassReader.java:425)

does not match current source lines.

**Comment by xtbe** [comment]
Hi,

Thanks for confirming. I got it fixed. Apparently some leftovers from a previous deployment were locking some of the libraries, this resulting in the old version.

I added it mainly because I couldn't find any feedback on Java 13. Also the user guide doesn't mention Java 13 in section 2.1 so could be good to add that.

https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/user-guide.html#d0e559

Report can now be closed.
